### ISSUE TITLE
Training Speed Tests

### ISSUE THREAD

**Issue by klannon** [klannon/deep_learning#5]
After doing #4, we should run the training on at least two of the "gpu" machines (1-5) and determine what the training speed per epoch is.  We should compare this to running with Torch.  We should also try the training with the GPU acceleration turned off to see how much speedup the GPU is providing.

**Comment by mdkdrnevich** [comment]
So this data is provided by pylearn2 in the log file, but should we write code to isolate the epoch speeds within the file or just manually scroll through and compare?

**Comment by klannon** [comment]
Since I think this is a number that we'll want to look at only 
intermittently (e.g. now, at the start of this process or later if we 
change something signficant), I don't think we need to write dedicated 
code.  We should just be able to extract the necessary information 
directly from the log files, which the help of grep, sed and awk.  (If 
you haven't played with those tools, I highly recommend taking the 
effort to pick them up!)

I would be interested if you could run the code in a few places a 
tabulate the results:  a few of the GPU machines using GPU or CPU, 
earth, using CPU (it doesn't have a GPU), and perhaps also your laptop.  
It would be good if we could understand whether there is any trend or 
pattern.  Can you also note some of the hardware details in this 
tabulation (CPU model and clock speed, memory, and GPU model, when 
relevant)?  For the GPU machines and earth, you can gather this 
information as follows:

CPU: cat /proc/cpuinfo
Memory: cat /proc/meminfo
GPU: cat /proc/driver/nvidia/gpus/*/information

**Comment by klannon** [comment]
Here is something that I could use some help tracking down.  I was reading about optimizing Theano performance here (http://deeplearning.net/software/theano/tutorial/using_gpu.html).  One of the first suggestions made is to use `allow_gc=False`.  I tried this by switching my `THEANO_FLAGS` from 

`setenv THEANO_FLAGS "blas.ldflags=-L/afs/crc.nd.edu/user/k/klannon/torchstuff/blas/install/lib -lopenblas,cuda.root=/afs/crc.nd.edu/x86_64_linux/c/cuda/7.0/lib,device=gpu,mode=FAST_RUN,floatX=float32"`

to 

`setenv THEANO_FLAGS "blas.ldflags=-L/afs/crc.nd.edu/user/k/klannon/torchstuff/blas/install/lib -lopenblas,cuda.root=/afs/crc.nd.edu/x86_64_linux/c/cuda/7.0/lib,device=gpu,mode=FAST_RUN,floatX=float32,allow_gc=False"`

But then the code crashed with the following error:

```
Data loaded: train_all_3v_ttbar_wjet (train)
Data loaded: train_all_3v_ttbar_wjet (valid)
Data loaded: test_all_3v_ttbar_wjet (test)
Data loaded: train_all_3v_ttbar_wjet (monitor)
Writing to /afs/crc.nd.edu/user/k/klannon/pylearn2/deep-learning/osu/run_3v_classification_accuracy.csv
Using=gpu
Writing to /afs/crc.nd.edu/user/k/klannon/pylearn2/deep-learning/osu/run_3v.log
Writing to /afs/crc.nd.edu/user/k/klannon/pylearn2/deep-learning/osu/run_3v.pkl
Traceback (most recent call last):
  File "run_3v.py", line 150, in <module>
    train(mytrain)
  File "run_3v.py", line 146, in train
    mytrain.main_loop()
  File "/afs/crc.nd.edu/user/k/klannon/pylearn2/pylearn2/pylearn2/train.py", line 201, in main_loop
    extension_continue = self.run_callbacks_and_monitoring()
  File "/afs/crc.nd.edu/user/k/klannon/pylearn2/pylearn2/pylearn2/train.py", line 259, in run_callbacks_and_monitoring
    extension.on_monitor(self.model, self.dataset, self.algorithm)
  File "/afs/crc.nd.edu/user/k/klannon/pylearn2/deep-learning/osu/accuracyTest.py", line 43, in on_monitor
    train_accuracy, valid_accuracy, test_accuracy = self.accuracy(self.args['train'].X, self.args['valid'].X, self.args['test'].X)
  File "/afs/crc.nd.edu/user/k/klannon/VE-deeplearning/lib/python2.7/site-packages/theano/compile/function_module.py", line 606, in __call__
    storage_map=self.fn.storage_map)
  File "/afs/crc.nd.edu/user/k/klannon/VE-deeplearning/lib/python2.7/site-packages/theano/compile/function_module.py", line 595, in __call__
    outputs = self.fn()
MemoryError: Error allocating 544000000 bytes of device memory (out of memory).
Apply node that caused the error: GpuElemwise{Composite{Cast{float32}(GT(i0, i1))},no_inplace}(GpuElemwise{Add}[(0, 0)].0, CudaNdarrayConstant{[[ 0.]]})
Inputs types: [CudaNdarrayType(float32, matrix), CudaNdarrayType(float32, (True, True))]
Inputs shapes: [(1360000, 100), (1, 1)]
Inputs strides: [(100, 1), (0, 0)]
Inputs values: ['not shown', <CudaNdarray object at 0x7f6edb38afb0>]

HINT: Re-running with most Theano optimization disabled could give you a back-trace of when this node was created. This can be done with by setting the Theano flag 'optimizer=fast_compile'. If that does not work, Theano opt
imizations can be disabled with 'optimizer=None'.
HINT: Use the Theano flag 'exception_verbosity=high' for a debugprint and storage map footprint of this apply node.
```

At first glance, it looks like this error is associated with `accuracyTest.py`.  Can someone take a look to determine whether we might be being a little careless with memory there?  My working hypothesis is that `allow_gc=False` disables Python garbage collection and if you're not careful, you can overrun memory my making too many accidental temporary copies of things, which Python won't take care of with garbage collection turned off.  Thoughts?

I should acknowledge that this could be a wild goose chase, and running with this flag might not change anything significant in the timing.

**Comment by klannon** [comment]
Update!

Playing around with settings, I was able to improve the training speed significantly under two scenarios.  All benchmarking done on gpu5:

Default:
- `setenv THEANO_FLAGS  "blas.ldflags=-L/afs/crc.nd.edu/user/k/klannon/torchstuff/blas/install/lib -lopenblas,cuda.root=/afs/crc.nd.edu/x86_64_linux/c/cuda/7.0/lib,device=gpu,mode=FAST_RUN,floatX=float32"`
- Performance:
  - Compiling sgd_update done. Time elapsed: 1.220503 seconds
  - compiling begin_record_entry done. Time elapsed: 1.435679 seconds
  - Compiling accum done. Time elapsed: 21.406867 seconds
  - Time this epoch: 0:02:21.975106

We can speed things up significantly on CPU only by doing the following:
- Set up openmpi and recompile OpenBLAS to use openmpi
  - `module load ompi/1.8.7-gcc-4.4.7`
  - Uncomment `USE_OPENMP=1` in Make.rules for OpenBLAS
  - Run `make`
- `setenv THEANO_FLAGS "blas.ldflags=-L/afs/crc.nd.edu/user/k/klannon/torchstuff/blas/install/lib -lopenblas,device=cpu,mode=FAST_RUN,floatX=float32"`
- `setenv OMP_NUM_THREADS 8`
- Change to `batch_size = 4096` in `run_3v.py`
- Performance:
  - Compiling sgd_update done. Time elapsed: 0.857060 seconds
  - compiling begin_record_entry done. Time elapsed: 1.198020 seconds
  - Compiling accum done. Time elapsed: 16.191721 seconds
  - Time this epoch: 16.031660 seconds

This setting will be good for using CPU when we don't have access to a GPU machine, such as if we want to run many jobs in the batch queue.  I stopped increasing the batch size and the number of threads once the performance plateaued.  I don't know if we could squeeze any more out by trying additinal combinations of batch sizes and number of threads.  Note gpu5 has only 12 cores so don't go above 12 threads there.  Also, don't go to earth and use a huge number of threads--that machine is shared with lots of people.

We can speed things up even more using a GPU with these settings:
- Comment out the `AccuracyMonitor` in the code to avoid the memory exhaustion crash above.  Note that in the standard log file output, there is a statistic called `XXX_y_misclass` which is 1-accuracy, with `XXX` = [train,test,valid].
- `setenv THEANO_FLAGS "blas.ldflags=-L/afs/crc.nd.edu/user/k/klannon/torchstuff/blas/install/lib -lopenblas,cuda.root=/afs/crc.nd.edu/x86_64_linux/c/cuda/7.0/lib,device=gpu,mode=FAST_RUN,floatX=float32,allow_gc=False"`
- `setenv OMP_NUM_THREADS 1`: 
- Change to `batch_size = 8192` in `run_3v.py`
- Performance:
  - Compiling sgd_update done. Time elapsed: 1.321658 seconds
  - compiling begin_record_entry done. Time elapsed: 1.466632 seconds
  - Compiling accum done. Time elapsed: 21.856220 seconds
  - Time this epoch: 2.502358 seconds

Again, I stopped messing around when the amount of improvement started to flatten out.  I'm not sure if there might be a setting at either higher or lower batch size that would perform better.  Adding more OMPI threads doesn't change performance.  Turning off the garbage collector improves the performance from 3.3 s to 2.5 s.  There might be other ways to try to squeeze performance out of the GPU, but I haven't had time to explore.  It might also be that some machines have a nicer GPU than gpu5 and we could go even faster there.

I'm going to start a run now and just let it go until tomorrow afternoon, just to see what happens.

**Comment by DayOfThePenguin** [comment]
Wow.  That's really fast.  The OSU folks are in for a shock tomorrow
afternoon.

**Comment by DayOfThePenguin** [comment]
I just started one on GPU2 that uses exponential decay (like the deep learning paper used).  My performance is about the same as (2.99 seconds per epoch).  This log file is going to be huge by the time I stop it tomorrow.

**Comment by mdkdrnevich** [comment]
That's an incredible improvement! Since y'all are running jobs I went ahead
and pushed the code I used to create the graphs I've been posting. Just
remember to load root before you try running it. Hopefully we get some good
results tomorrow!

**Comment by klannon** [comment]
@DayOfThePenguin:  Any luck with your big run?
My messed up my big run.  (This is what I get for trying to do it right before going to bed while feeling ill...)  But in playing around today, I'm starting to suspect that time per epoch really isn't the best metric for how quickly we're training.  This is because we make progress in training every _batch_ (e.g. the network weights are updated every batch).  So, just because you've run a bunch of epochs doesn't mean that you've really made much progress.  It depends on how many times you've updated your weights (i.e. how many batches you fit in).  If you look at it from this perspective, then `batch_size = 16` might be the best one.  The epochs take a long time (I've gotten them down to 1 minute on the GPU thanks to turning off GC), but you do many batches (i.e. update your weights many times).  I'm trying to check now to see which is more important, the total number of epochs or the total number of batches.

**Comment by klannon** [comment]
We should look at this page to see if any of these techniques can be used to speed up the GPU processing:

http://deeplearning.net/software/theano/tutorial/aliasing.html

**Comment by klannon** [comment]
Another mystery:  Is training time really the full time for the epoch, or is there something else that's taking up a bunch of time each epoch.  Noticed this in the log file output:

```
Time this epoch: 18.888517 seconds
    total_seconds_last_epoch: 76.4794082642
    training_seconds_this_epoch: 18.8883342743
```

If I take this at face value, it seems like the total epoch time is 76 s of which, only 19 s are spent in training.  Does that seem right?  Any ideas what's eating up the rest of the time?

**Comment by klannon** [comment]
In this thread (https://groups.google.com/forum/#!msg/pylearn-users/1qfdO92WhBQ/tvsJSSG4HjkJ) it is suggested that the `total_seconds_last_epoch` statistic includes mainly monitoring time.  If that's true, our monitoring is going really slowly.  We should look into this.

**Comment by mdkdrnevich** [comment]
How important are the monitoring statistics? Because if we are only looking at classification accuracy then it might be beneficial if I try to make AccuracyMonitor more memory efficient so we can use that, since it only takes the algorithm a few seconds to determine the accuracy (based on manually counting the time).

**Comment by DayOfThePenguin** [comment]
Couldn't we also just try copying the pylearn implementation of particular
monitoring channels into our own code?

**Comment by mdkdrnevich** [comment]
Actually, it wouldn't even be necessary to improve the memory management of the accuracy file. I just did a test run on my laptop with allow_gc=False and monitoring turned off but with AccuracyMonitor on, and my laptop never experienced a MemoryError. What I did notice is that at the end of each epoch there would be a spike of about 1.5GB in memory and then would go back down. From what I can tell it looks like most of the memory inefficiency must be in the monitoring part then if it can't spare 1.5GB of ram when it gets to AccuracyMonitor

**Comment by mdkdrnevich** [comment]
Or what Colin said

**Comment by klannon** [comment]
I'm wondering if there isn't something else besides monitoring that can account for the excess time.  I guess I should try to double check the timing information we have here, but if I look at different `batch_size` values the ratio of total/training seconds seems consistently around 4.  This doesn't make sense to me.  I would think that the monitoring would happen once per epoch and not once per batch.  If that's true, then there would be no reason for the monitoring time to change at all with `batch_size`.  I guess it would be helpful to do some poking to understand what all is going on during one epoch.

**Comment by DayOfThePenguin** [comment]
This might help explain why we got to much faster when we changed the batch size to 8192:

https://groups.google.com/forum/#!topic/theano-users/5AhLuUqDga0

Essentially, during one epoch, the train data will get split up into `X` batches of `batch_size` (where `X=len(train_data) / batch_size` ), so if we make the batch size big, we just end up updating the weights on the edges fewer times per epoch (because an epoch is one run through the entire training set) than if the batch size is small.  I think there are some lower-level hardware things going on (e.g. whether certain data are still in the cache) that make using larger batches more computationally efficient, so the batch size might be something else we need to play with to see what gives us the shortest time to update the weights once (which would be `training_time / X` ).

And just a theory about what @klannon mentioned above: 

Maybe monitoring is performed once a batch.  Though this might not seem right to us, it might explain why the monitoring data listed for each epoch looks like this:

test_r1_mean_x_max_u
test_r1_mean_x_mean_u
test_r1_mean_x_min_u 

The ..._max_u, ..._mean_u, and ..._min_u could be the maximum, minimum, and average value of the channel that was recorded across all the batches that were analyzed during this epoch.
